File: python/06-memoria/03-episodica/episodic_decay.py
"""
Memoria episódica con decay exponencial, refuerzo y ciclo de vida.

Requiere: sqlite-vec (pip install sqlite-vec), sentence-transformers o cualquier
función de embedding. El ejemplo usa un vector ficticio para que el código corra
sin dependencias externas — reemplaza `_embed` con tu modelo real.

Cómo ejecutar:
    make py SCRIPT=python/06-memoria/03-episodica/episodic_decay.py

Qué esperar:
    Episodios que decaen con el tiempo y se refuerzan al ser recuperados.
    Similaridad coseno + decay exponencial para scoring de relevancia.
    Nota: usa vectores ficticios — reemplaza _embed() con tu modelo real.

Variables de entorno:
    MODEL — modelo a usar (default: claude-sonnet-4-6)
"""
from __future__ import annotations

import math
import sqlite3
import time
import uuid
from dataclasses import dataclass, field
from typing import Optional

# ---------------------------------------------------------------------------
# Esquema
# ---------------------------------------------------------------------------

CREATE_SQL = """
CREATE TABLE IF NOT EXISTS episodios (
    id             TEXT PRIMARY KEY,
    contenido      TEXT NOT NULL,
    timestamp      REAL NOT NULL,
    sesion_id      TEXT,
    fuerza         REAL NOT NULL DEFAULT 1.0,
    half_life_dias REAL NOT NULL DEFAULT 7.0,
    accesos        INTEGER NOT NULL DEFAULT 0,
    ultimo_acceso  REAL NOT NULL,
    estado         TEXT NOT NULL DEFAULT 'activo',  -- activo|olvidado|consolidado|tombstone
    reemplaza_a    TEXT REFERENCES episodios(id)
);
CREATE INDEX IF NOT EXISTS idx_episodios_estado   ON episodios(estado);
CREATE INDEX IF NOT EXISTS idx_episodios_fuerza   ON episodios(fuerza);
CREATE INDEX IF NOT EXISTS idx_episodios_timestamp ON episodios(timestamp);
"""

UMBRAL_OLVIDO    = 0.05   # fuerza mínima antes de marcar como olvidado
DELTA_REFUERZO   = 0.15   # incremento de fuerza por cada recuperación con uso
UMBRAL_CONSOLIDAR = 0.85  # similitud coseno para agrupar episodios en cluster


@dataclass
class Episodio:
    id:             str
    contenido:      str
    timestamp:      float
    sesion_id:      Optional[str]
    fuerza:         float = 1.0
    half_life_dias: float = 7.0
    accesos:        int   = 0
    ultimo_acceso:  float = field(default_factory=time.time)
    estado:         str   = "activo"
    reemplaza_a:    Optional[str] = None


# ---------------------------------------------------------------------------
# Función de embedding (placeholder — reemplazar con modelo real)
# ---------------------------------------------------------------------------

def _embed(texto: str) -> list[float]:
    """
    Embedding ficticio de dimensión 4 para demostración.
    En producción: sentence_transformers, openai.embeddings.create, etc.
    """
    words = texto.lower().split()
    vec = [0.0, 0.0, 0.0, 0.0]
    for i, w in enumerate(words[:4]):
        vec[i % 4] += hash(w) % 100 / 100.0
    norm = math.sqrt(sum(x * x for x in vec)) or 1.0
    return [x / norm for x in vec]


def _coseno(a: list[float], b: list[float]) -> float:
    dot  = sum(x * y for x, y in zip(a, b))
    na   = math.sqrt(sum(x * x for x in a)) or 1e-9
    nb   = math.sqrt(sum(x * x for x in b)) or 1e-9
    return dot / (na * nb)


# ---------------------------------------------------------------------------
# AlmacenEpisodico
# ---------------------------------------------------------------------------

class AlmacenEpisodico:
    def __init__(self, db_path: str = ":memory:"):
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.row_factory = sqlite3.Row
        self.conn.executescript(CREATE_SQL)
        self.conn.commit()
        # Caché de embeddings para búsqueda sin sqlite-vec
        self._embeddings: dict[str, list[float]] = {}

    # ------------------------------------------------------------------
    # record — insertar nuevo episodio
    # ------------------------------------------------------------------

    def record(
        self,
        contenido: str,
        sesion_id: Optional[str] = None,
        half_life_dias: float = 7.0,
    ) -> Episodio:
        ep = Episodio(
            id=str(uuid.uuid4()),
            contenido=contenido,
            timestamp=time.time(),
            sesion_id=sesion_id,
            half_life_dias=half_life_dias,
            ultimo_acceso=time.time(),
        )
        self.conn.execute(
            """INSERT INTO episodios
               (id, contenido, timestamp, sesion_id, fuerza, half_life_dias,
                accesos, ultimo_acceso, estado)
               VALUES (?,?,?,?,?,?,?,?,?)""",
            (ep.id, ep.contenido, ep.timestamp, ep.sesion_id,
             ep.fuerza, ep.half_life_dias, ep.accesos, ep.ultimo_acceso, ep.estado),
        )
        self.conn.commit()
        self._embeddings[ep.id] = _embed(contenido)
        return ep

    # ------------------------------------------------------------------
    # recall — recuperar episodios relevantes con scoring multi-señal
    # ------------------------------------------------------------------

    def recall(
        self,
        query: str,
        top_k: int = 5,
        skip_reinforce: bool = False,
    ) -> list[tuple[Episodio, float]]:
        """
        Scoring = 0.5 × similitud_coseno + 0.3 × recencia + 0.2 × fuerza_actual

        skip_reinforce=True para consultas exploratorias que no deben
        distorsionar las estadísticas de uso.
        """
        q_vec = _embed(query)
        ahora = time.time()

        rows = self.conn.execute(
            "SELECT * FROM episodios WHERE estado = 'activo'"
        ).fetchall()

        if not rows:
            return []

        # Calcular score para cada episodio activo
        candidatos: list[tuple[Episodio, float]] = []
        timestamps = [r["timestamp"] for r in rows]
        t_min, t_max = min(timestamps), max(timestamps)
        t_rango = (t_max - t_min) or 1.0

        for r in rows:
            ep_id = r["id"]
            ep_vec = self._embeddings.get(ep_id, _embed(r["contenido"]))

            sim       = max(0.0, _coseno(q_vec, ep_vec))
            recencia  = (r["timestamp"] - t_min) / t_rango          # 0..1
            fuerza_n  = min(1.0, r["fuerza"])                        # 0..1

            score = 0.5 * sim + 0.3 * recencia + 0.2 * fuerza_n

            ep = Episodio(
                id=ep_id, contenido=r["contenido"],
                timestamp=r["timestamp"], sesion_id=r["sesion_id"],
                fuerza=r["fuerza"], half_life_dias=r["half_life_dias"],
                accesos=r["accesos"], ultimo_acceso=r["ultimo_acceso"],
                estado=r["estado"],
            )
            candidatos.append((ep, score))

        candidatos.sort(key=lambda x: x[1], reverse=True)
        resultado = candidatos[:top_k]

        # Refuerzo — solo cuando el episodio contribuye a una respuesta real
        if not skip_reinforce:
            for ep, _ in resultado:
                self.conn.execute(
                    """UPDATE episodios
                       SET fuerza = MIN(fuerza + ?, 2.0),
                           accesos = accesos + 1,
                           ultimo_acceso = ?
                       WHERE id = ?""",
                    (DELTA_REFUERZO, ahora, ep.id),
                )
            self.conn.commit()

        return resultado

    # ------------------------------------------------------------------
    # tick_lifecycle — aplicar decay y consolidar; ejecutar en background
    # ------------------------------------------------------------------

    def tick_lifecycle(self) -> dict[str, int]:
        """
        Aplica decay exponencial a todos los episodios activos.
        Devuelve un resumen: cuántos olvidados y cuántos consolidados.
        """
        ahora = time.time()
        stats = {"olvidados": 0, "consolidados": 0}

        rows = self.conn.execute(
            "SELECT * FROM episodios WHERE estado = 'activo'"
        ).fetchall()

        for r in rows:
            elapsed_dias = (ahora - r["ultimo_acceso"]) / 86400.0
            nueva_fuerza = r["fuerza"] * math.exp(
                -math.log(2) * elapsed_dias / r["half_life_dias"]
            )

            if nueva_fuerza < UMBRAL_OLVIDO:
                self.conn.execute(
                    "UPDATE episodios SET fuerza=?, estado='olvidado' WHERE id=?",
                    (nueva_fuerza, r["id"]),
                )
                stats["olvidados"] += 1
            else:
                self.conn.execute(
                    "UPDATE episodios SET fuerza=? WHERE id=?",
                    (nueva_fuerza, r["id"]),
                )

        self.conn.commit()

        # Consolidación: agrupar episodios similares recientes en un resumen
        ventana_segundos = 24 * 3600
        recientes = self.conn.execute(
            "SELECT * FROM episodios WHERE estado='activo' AND timestamp > ?",
            (ahora - ventana_segundos,),
        ).fetchall()

        # Clustering por similitud coseno (naive O(n²) — suficiente para <200 eps.)
        visitados: set[str] = set()
        for r in recientes:
            if r["id"] in visitados:
                continue
            cluster = [r]
            vec_r = self._embeddings.get(r["id"], _embed(r["contenido"]))
            for other in recientes:
                if other["id"] == r["id"] or other["id"] in visitados:
                    continue
                vec_o = self._embeddings.get(other["id"], _embed(other["contenido"]))
                if _coseno(vec_r, vec_o) >= UMBRAL_CONSOLIDAR:
                    cluster.append(other)
                    visitados.add(other["id"])
            visitados.add(r["id"])

            if len(cluster) >= 3:
                self._consolidar_cluster(cluster)
                stats["consolidados"] += len(cluster)

        return stats

    def _consolidar_cluster(self, cluster: list[sqlite3.Row]) -> None:
        """Fusiona un cluster de episodios similares en uno solo."""
        textos   = [r["contenido"] for r in cluster]
        resumen  = f"[Consolidado de {len(cluster)} episodios]\n" + " | ".join(textos[:3])
        fuerza   = max(r["fuerza"] for r in cluster)
        half_life = cluster[0]["half_life_dias"]

        nuevo = self.record(resumen, half_life_dias=half_life)
        self.conn.execute(
            "UPDATE episodios SET fuerza=? WHERE id=?",
            (fuerza, nuevo.id),
        )
        ids = [r["id"] for r in cluster]
        self.conn.execute(
            f"UPDATE episodios SET estado='consolidado' WHERE id IN ({','.join('?'*len(ids))})",
            ids,
        )
        self.conn.commit()


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    store = AlmacenEpisodico()

    store.record("El usuario prefiere respuestas concisas", half_life_dias=30)
    store.record("Bug en auth.py línea 247: condición invertida", half_life_dias=3)
    store.record("Decidimos usar PostgreSQL en lugar de SQLite para producción", half_life_dias=90)
    store.record("El módulo de billing tiene deuda técnica: lógica duplicada", half_life_dias=7)

    print("--- recall: 'base de datos producción' ---")
    for ep, score in store.recall("base de datos producción", top_k=3):
        print(f"  [{score:.3f}] {ep.contenido[:60]}")

    print("\n--- recall exploratorio (skip_reinforce) ---")
    for ep, score in store.recall("preferencias usuario", top_k=2, skip_reinforce=True):
        print(f"  [{score:.3f}] {ep.contenido[:60]}")

    print("\n--- tick_lifecycle ---")
    stats = store.tick_lifecycle()
    print(f"  olvidados: {stats['olvidados']}, consolidados: {stats['consolidados']}")

File: typescript/06-memoria/03-episodica/episodic_decay.ts
// Memoria episódica con decay exponencial, refuerzo y ciclo de vida.
// Usa better-sqlite3 para persistencia. El embedding usa un vector ficticio
// de dimensión 4 — reemplazar con modelo real en producción.

// Cómo ejecutar: make ts SCRIPT=typescript/06-memoria/03-episodica/episodic_decay.ts


import { Database } from "bun:sqlite";

const CREATE_SQL = `
CREATE TABLE IF NOT EXISTS episodios (
    id             TEXT PRIMARY KEY,
    contenido      TEXT NOT NULL,
    timestamp      REAL NOT NULL,
    sesion_id      TEXT,
    fuerza         REAL NOT NULL DEFAULT 1.0,
    half_life_dias REAL NOT NULL DEFAULT 7.0,
    accesos        INTEGER NOT NULL DEFAULT 0,
    ultimo_acceso  REAL NOT NULL,
    estado         TEXT NOT NULL DEFAULT 'activo'
);
CREATE INDEX IF NOT EXISTS idx_episodios_estado    ON episodios(estado);
CREATE INDEX IF NOT EXISTS idx_episodios_fuerza    ON episodios(fuerza);
CREATE INDEX IF NOT EXISTS idx_episodios_timestamp ON episodios(timestamp);
`;

const UMBRAL_OLVIDO = 0.05;
const DELTA_REFUERZO = 0.15;
const UMBRAL_CONSOLIDAR = 0.85;

interface Episodio {
  id: string;
  contenido: string;
  timestamp: number;
  sesionId: string | null;
  fuerza: number;
  halfLifeDias: number;
  accesos: number;
  ultimoAcceso: number;
  estado: string;
}

function generateId(): string {
  return Math.random().toString(36).slice(2) + Date.now().toString(36);
}

function embed(texto: string): number[] {
  const words = texto.toLowerCase().split(/\s+/);
  const vec = [0.0, 0.0, 0.0, 0.0];
  for (let i = 0; i < Math.min(words.length, 4); i++) {
    let h = 0;
    for (const c of words[i]) {
      h = (Math.imul(31, h) + c.charCodeAt(0)) | 0;
    }
    vec[i % 4] += (Math.abs(h) % 100) / 100.0;
  }
  const norm = Math.sqrt(vec.reduce((s, x) => s + x * x, 0)) || 1.0;
  return vec.map((x) => x / norm);
}

function coseno(a: number[], b: number[]): number {
  const dot = a.reduce((s, x, i) => s + x * b[i], 0);
  const na = Math.sqrt(a.reduce((s, x) => s + x * x, 0)) || 1e-9;
  const nb = Math.sqrt(b.reduce((s, x) => s + x * x, 0)) || 1e-9;
  return dot / (na * nb);
}

class AlmacenEpisodico {
  private db: Database;
  private embeddings: Map<string, number[]> = new Map();

  constructor(dbPath: string = ":memory:") {
    this.db = new Database(dbPath);
    this.db.exec(CREATE_SQL);
  }

  record(contenido: string, sesionId: string | null = null, halfLifeDias: number = 7.0): Episodio {
    const id = generateId();
    const now = Date.now() / 1000;
    this.db
      .prepare(
        `INSERT INTO episodios (id, contenido, timestamp, sesion_id, fuerza, half_life_dias, accesos, ultimo_acceso, estado)
         VALUES (?,?,?,?,1.0,?,0,?,'activo')`,
      )
      .run(id, contenido, now, sesionId, halfLifeDias, now);
    this.embeddings.set(id, embed(contenido));
    return { id, contenido, timestamp: now, sesionId, fuerza: 1.0, halfLifeDias, accesos: 0, ultimoAcceso: now, estado: "activo" };
  }

  recall(query: string, topK: number = 5, skipReinforce: boolean = false): [Episodio, number][] {
    const qVec = embed(query);
    const ahora = Date.now() / 1000;

    const rows = this.db.prepare("SELECT * FROM episodios WHERE estado = 'activo'").all() as Record<string, unknown>[];
    if (!rows.length) return [];

    const timestamps = rows.map((r) => r.timestamp as number);
    const tMin = Math.min(...timestamps);
    const tMax = Math.max(...timestamps);
    const tRango = (tMax - tMin) || 1.0;

    const candidatos: [Episodio, number][] = rows.map((r) => {
      const epId = r.id as string;
      const epVec = this.embeddings.get(epId) ?? embed(r.contenido as string);
      const sim = Math.max(0, coseno(qVec, epVec));
      const recencia = ((r.timestamp as number) - tMin) / tRango;
      const fuerzaN = Math.min(1.0, r.fuerza as number);
      const score = 0.5 * sim + 0.3 * recencia + 0.2 * fuerzaN;
      const ep: Episodio = {
        id: epId, contenido: r.contenido as string, timestamp: r.timestamp as number,
        sesionId: r.sesion_id as string | null, fuerza: r.fuerza as number,
        halfLifeDias: r.half_life_dias as number, accesos: r.accesos as number,
        ultimoAcceso: r.ultimo_acceso as number, estado: r.estado as string,
      };
      return [ep, score];
    });

    candidatos.sort((a, b) => b[1] - a[1]);
    const resultado = candidatos.slice(0, topK);

    if (!skipReinforce) {
      for (const [ep] of resultado) {
        this.db
          .prepare(
            `UPDATE episodios SET fuerza = MIN(fuerza + ?, 2.0), accesos = accesos + 1, ultimo_acceso = ? WHERE id = ?`,
          )
          .run(DELTA_REFUERZO, ahora, ep.id);
      }
    }
    return resultado;
  }

  tickLifecycle(): { olvidados: number; consolidados: number } {
    const ahora = Date.now() / 1000;
    const stats = { olvidados: 0, consolidados: 0 };

    const rows = this.db.prepare("SELECT * FROM episodios WHERE estado = 'activo'").all() as Record<string, unknown>[];
    for (const r of rows) {
      const elapsedDias = (ahora - (r.ultimo_acceso as number)) / 86400.0;
      const nuevaFuerza = (r.fuerza as number) * Math.exp((-Math.LN2 * elapsedDias) / (r.half_life_dias as number));
      if (nuevaFuerza < UMBRAL_OLVIDO) {
        this.db.prepare("UPDATE episodios SET fuerza=?, estado='olvidado' WHERE id=?").run(nuevaFuerza, r.id);
        stats.olvidados++;
      } else {
        this.db.prepare("UPDATE episodios SET fuerza=? WHERE id=?").run(nuevaFuerza, r.id);
      }
    }

    const ventanaSegundos = 24 * 3600;
    const recientes = this.db
      .prepare("SELECT * FROM episodios WHERE estado='activo' AND timestamp > ?")
      .all(ahora - ventanaSegundos) as Record<string, unknown>[];

    const visitados = new Set<string>();
    for (const r of recientes) {
      if (visitados.has(r.id as string)) continue;
      const cluster: Record<string, unknown>[] = [r];
      const vecR = this.embeddings.get(r.id as string) ?? embed(r.contenido as string);
      for (const other of recientes) {
        if (other.id === r.id || visitados.has(other.id as string)) continue;
        const vecO = this.embeddings.get(other.id as string) ?? embed(other.contenido as string);
        if (coseno(vecR, vecO) >= UMBRAL_CONSOLIDAR) {
          cluster.push(other);
          visitados.add(other.id as string);
        }
      }
      visitados.add(r.id as string);
      if (cluster.length >= 3) {
        this.consolidarCluster(cluster);
        stats.consolidados += cluster.length;
      }
    }
    return stats;
  }

  private consolidarCluster(cluster: Record<string, unknown>[]): void {
    const textos = cluster.map((r) => r.contenido as string);
    const resumen = `[Consolidado de ${cluster.length} episodios]\n` + textos.slice(0, 3).join(" | ");
    const fuerza = Math.max(...cluster.map((r) => r.fuerza as number));
    const halfLife = cluster[0].half_life_dias as number;
    const nuevo = this.record(resumen, null, halfLife);
    this.db.prepare("UPDATE episodios SET fuerza=? WHERE id=?").run(fuerza, nuevo.id);
    const ids = cluster.map((r) => r.id as string);
    this.db
      .prepare(`UPDATE episodios SET estado='consolidado' WHERE id IN (${ids.map(() => "?").join(",")})`)
      .run(...ids);
  }
}

const store = new AlmacenEpisodico();
store.record("El usuario prefiere respuestas concisas", null, 30);
store.record("Bug en auth.py línea 247: condición invertida", null, 3);
store.record("Decidimos usar PostgreSQL en lugar de SQLite para producción", null, 90);
store.record("El módulo de billing tiene deuda técnica: lógica duplicada", null, 7);

console.log("--- recall: 'base de datos producción' ---");
for (const [ep, score] of store.recall("base de datos producción", 3)) {
  console.log(`  [${score.toFixed(3)}] ${ep.contenido.slice(0, 60)}`);
}

console.log("\n--- recall exploratorio (skipReinforce) ---");
for (const [ep, score] of store.recall("preferencias usuario", 2, true)) {
  console.log(`  [${score.toFixed(3)}] ${ep.contenido.slice(0, 60)}`);
}

console.log("\n--- tick_lifecycle ---");
const stats = store.tickLifecycle();
console.log(`  olvidados: ${stats.olvidados}, consolidados: ${stats.consolidados}`);
